File: legacy/src/logic/solvers/LeapfrogSolver.java
package logic.solvers;

import model.physics.Body2D;
import model.physics.Universe2D;

public 
class LeapfrogSolver 
extends Physics2DSolver
{
	public 
	LeapfrogSolver(Universe2D u) 
	{
		super(u);
		timestep = 0.1;
		displayName = "Leapfrog";
	}
	
	@Override
	public Physics2DSolver getNewInstance(Universe2D u) {
		return new LeapfrogSolver(u);
	}

	@Override
	public void step() {
		computeAllBodiesAccelerations();

		for(Body2D b1 : universe.bodies)
		{
			if(timeElapsed == 0)
				b1.velocity.add(b1.acceleration.times(timestep/2));
			else
				b1.velocity.add(b1.acceleration.times(timestep));
			
			b1.position.add(b1.velocity.times(timestep));
		}
		
		timeElapsed += timestep;
	}

}

File: src/physics/solvers/leapfrog_solver.cpp
/*
 * leapfrog_solver.cpp
 *
 *  Created on: 12 de jul de 2016
 *      Author: carlosfaruolo
 */

#include "leapfrog_solver.hpp"

#include <vector>
using std::vector;

DEFINE_CLASS_FACTORY(LeapfrogSolver, "Leapfrog");

LeapfrogSolver::LeapfrogSolver(Universe2D& u)
: AbstractPhysics2DSolver(&CLASS_FACTORY, u, 0.1)
{}

void LeapfrogSolver::step()
{
	computeAccelerations();

	foreach(Body2D*, body, vector<Body2D*>, universe.bodies)
	{
		if(timeElapsed == 0)
			body->velocity += body->acceleration * (timestep*0.5);
		else
			body->velocity += body->acceleration * timestep;

		body->position += body->velocity * timestep;
	}

	timeElapsed += timestep;
}
